core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2014–2019 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.destination.s3.impl
+
+import quasar.destination.s3.{Bucket, ObjectKey, Upload}
+
+import slamdata.Predef._
+
+import cats.effect.{Concurrent, ExitCase}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.Stream
+import monix.catnap.syntax._
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.core.async.AsyncRequestBody
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.{
+  AbortMultipartUploadRequest,
+  AbortMultipartUploadResponse,
+  CompleteMultipartUploadRequest,
+  CompleteMultipartUploadResponse,
+  CompletedMultipartUpload,
+  CompletedPart,
+  CreateMultipartUploadRequest,
+  CreateMultipartUploadResponse,
+  UploadPartRequest
+}
+
+final case class DefaultUpload[F[_]: Concurrent](client: S3AsyncClient, partSize: Int) extends Upload[F] {
```

Comment:
The fact that there's only a `Concurrent` bound here indicates that the thread shifting is not being done correctly. In other words, the thread pools inside of the AWS client are leaking into sdbe. In order to fix this, a `ContextShift` is required (note the difference between `IO.fromFuture` in Cats Effect 1.0 and our `fromFutureShift` function).